### PR TITLE
Prohibit usage of retired names in semantic conventions

### PR DESCRIPTION
**Pull request by tigrannajaryan**
This change adds a prohibition clause that requires that no old
metric or attribute name is used for a new attribute.

This is important to ensure reversibility of schema transformation
(converting from a new version to an old version of schema).

Without this restriction the following is possible:

Schema version 1. Attribute A exists.
Schema version 2. Attribute A is renamed to B. Appropriate schema file is created.
Schema version 3. Attribute A is introduced (a completely different new attribute).

Now attempting to go from Version 3 to version 1 is impossible since it requires
renaming B to A (for the change in version 2), but a different attribute A already exists.

### CODE REVIEW

**Comment by Oberon00** [comment]
Is it expected that lossless conversion from new to old versions works? That does not seem like a feasible thing to guarantee to me.

**Comment by yurishkuro** [comment]
> Is it expected that lossless conversion from new to old versions works? That does not seem like a feasible thing to guarantee to me.

I agree, at least there needs to be some grace period after which earlier versions of the schema should be considered unreachable and the old name could be used again.

**Comment by tigrannajaryan** [comment]
I agree that we probably want to have a retirement policy longer term. But until we have it this is a relatively cheap way to ensure backward conversions to old versions work.

Backward conversions are very important for backends. The backend often will understand a specific telemetry version. When Otel releases a new version the backend cannot be expected to immediately be updated and start understanding that new version. This means backwards conversions are going to be fairly typical and I think if we we can ensure they are guaranteed to work we should do it. I think this particular limitation is worth it.

Besides, it is generally a good practice to follow to avoid confusion. Again, I agree that over long periods of time likely you want to have a different retirement policy, e.g. something that is unused for years can be probably reused. Once we have that policy we can rethink the reuse prohibition rule.

**Review by iNikem — APPROVED** [review]

File: specification/common/attribute-naming.md [hunk 1]
```diff
@@ -67,6 +68,16 @@ Names SHOULD follow these rules:
   [Metric Name Pluralization Guidelines](../metrics/semantic_conventions/README.md#pluralization)
   SHOULD be followed for the attribute name.
 
+## Name Reuse Prohibition
+
+It is prohibited to introduce a new attribute with a name that matches a name of
```
Comment (SergeyKanzhelev) [inline specification/common/attribute-naming.md:73]: Can we use MUST NOT or SHOULD NOT language here instead of prohibited.

Comment (MrAlias) [reply specification/common/attribute-naming.md:73]:
```suggestion
A new attribute MUST NOT be added with the same name as
```

**Review by Oberon00 — APPROVED** [review]
@arminru @dyladan, others at Dynatrace and me discussed this PR shortly. We do not really agree with the motivation, or rather, we do not have that use case yet. But we think that adding this restriction does no harm and is beneficial also for humans to avoid confusing scenarios.

**Review by jmacd — APPROVED** [review]

File: specification/common/attribute-naming.md (hunk 1 above)
Comment (tigrannajaryan) [reply specification/common/attribute-naming.md]: Fixed.

File: specification/metrics/semantic_conventions/README.md [hunk 2]
```diff
@@ -66,6 +67,16 @@ and confusion for end users. (For example, prefer `process.runtime.java.gc*` ove
 `process.runtime.gc.*`.) Measures of many operating system metrics are similarly
 ambiguous.
 
+### Name Reuse Prohibition
+
+It is prohibited to introduce a new metric with a name that matches a name of
+a metric that existed in the past but was renamed (with a corresponding
+schema file).
```
Comment (MrAlias) [inline specification/metrics/semantic_conventions/README.md:74]: Can we copy the same wording from specification/common/attribute-naming.md here and in specification/trace/semantic_conventions/README.md?

Comment (tigrannajaryan) [reply specification/metrics/semantic_conventions/README.md]: Copied.

**Review by SergeyKanzhelev — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

File: specification/common/attribute-naming.md [hunk 3]
```diff
@@ -8,6 +8,7 @@
 <!-- toc -->
 
 - [Name Pluralization guidelines](#name-pluralization-guidelines)
+- [Name Reuse Prohibition](#name-reuse-prohibition)
```
Comment (arminru) [inline specification/common/attribute-naming.md:11]: Changelog please :-)

**Review by arminru — APPROVED** [review]

File: specification/common/attribute-naming.md (hunk 3 above)
Comment (tigrannajaryan) [reply specification/common/attribute-naming.md:11]: Added.

**Pull request merged**